### ISSUE TITLE
Dark theme leaves light GTK title bar

### ISSUE THREAD

**Issue by Yajo** [vector-im/element-web#15997]
### Description

Dark theme looks ugly on Linux (specifically, under GNOME; I don't know on other desktops)

### Steps to reproduce

- Use dark theme.

Logs being sent: no

![imagen](https://user-images.githubusercontent.com/973709/102618706-2bdd6800-4133-11eb-9596-70064c70e835.png)

### Version information

- **Platform**: desktop

For the desktop app:

- **OS**: Fedora Silverblue 33
- **Version**: 1.7.15

**Comment by j-james** [comment]
FWIW, I'm not seeing this behavior under XFCE. Seems like a GNOME-specific bug (perhaps even a GNOME bug) - normal GTK also doesn't allow clients to change the title bars.

**Comment by Yajo** [comment]
AFAIK it's an option for GTK apps. See https://stackoverflow.com/questions/15997464/how-do-i-make-my-gtk-app-use-the-dark-theme

**Comment by SimonBrandner** [comment]
@Yajo I am afraid we can't do anything about this. As you said [it is possible to change the title bar colour in GTK apps](https://stackoverflow.com/questions/15997464/how-do-i-make-my-gtk-app-use-the-dark-theme) through `C`. 

The problem is that Element is built using Electron, which is one "layer" above that. We would have to be able to set the title bar colour in `Javascript` and that [doesn't seem to be possible](https://stackoverflow.com/questions/46222035/set-color-of-windows-titlebar-electron-js/47739033).

**Comment by Yajo** [comment]
I'm probably missing something, but https://github.com/electron/electron/commit/058d0bbe19a61b0709ac75cb258ee65147d9ac57 and https://www.electronjs.org/docs/tutorial/dark-mode suggest that this is supported.

AFAIK, Spotify is an electron app on Linux and has dark native borders.

**Comment by SimonBrandner** [comment]
@Yajo It seems you're right. I am sorry for my mistake of trusting the first thing Google showed me. I'll try to look into this

**Comment by SimonBrandner** [comment]
@Yajo I've tested it and sadly it's not looking good. The `nativeTheme.themeSource` seems to only affect the dev tools. It has no effect on the title bar or the file picking dialog. 

Maybe I am missing something. I am not running Gnome on my machine. Therefore I tried this in a VM with live USB of Fedora and it uses Wayland. I don't know if that has any influence, but I am going to try with Xorg later.

Screenshot:

![Screenshot_20201222_192854](https://user-images.githubusercontent.com/25768714/102920993-f070cf80-448b-11eb-9ba3-e45c56640769.png)

Code:

```javascript
const { app, BrowserWindow, nativeTheme } = require('electron')

function createWindow () {
  const win = new BrowserWindow({
    width: 800,
    height: 600,
    webPreferences: {
      nodeIntegration: true
    }
  })

  win.loadURL("https://app.element.io");
}

app.whenReady().then(() => {
    createWindow();
    nativeTheme.themeSource = "dark";
});

app.on('window-all-closed', () => {
  if (process.platform !== 'darwin') {
    app.quit()
  }
})

app.on('activate', () => {
  if (BrowserWindow.getAllWindows().length === 0) {
    createWindow()
  }
})
```

**Comment by SimonBrandner** [comment]
@Yajo I've tried with `Manjaro Gnome`. It had some UI tweaks so in `Gnome Tweaks` > `Appearance` I set everything to default and then uninstalled `Gnome Tweaks`. `nativeTheme.themeSource` still seems to not affect the title bar. There is one weird thing the title bar didn't really look like `Adwaita` for either colour-scheme.

I am going to take a break from this today. I would like to try this out once more with clean Gnome. Although I am not sure which distro to use. Maybe Debian has live ISOs, I will take a look. If that won't work and we can't spot a problem in the code I am afraid this isn't possible

**Comment by Yajo** [comment]
I've been reading a bunch of electron issues all around and it's crazy but seems almost impossible unless you use a package that renders a header bar and hide the native one... https://github.com/electron/electron/issues/11907 may help in the future.

Fedora has almost 100% upstream gnome and live bootable images. Remember to log out and back in in x11 mode.

**Comment by SimonBrandner** [comment]
@Yajo You mentioned Spotify before. Does it have a native title bar?

**Comment by Yajo** [comment]
Yes, at least it looks completely native: 
![image](https://user-images.githubusercontent.com/973709/102964505-c1864800-44e3-11eb-9154-1c15aa5dece2.png)

**Comment by SimonBrandner** [comment]
@Yajo I've tried with `Debian Gnome` and got the same results... I've also tried to install Spotify and it has had the `light` title bar as well.

I think I've run out of ideas. If you have another idea, I can try. But I think we sadly can't do anything about this for now.

**Comment by Yajo** [comment]
> I've also tried to install Spotify and it has had the `light` title bar as well.

Oh really? Where did you install it from? I use this package https://www.flathub.org/apps/details/com.spotify.Client

In Fedora 33 it displays nicely dark as you see in https://github.com/vector-im/element-web/issues/15997#issuecomment-749952680

**Comment by Yajo** [comment]
I'm chatting with the gnome design team (you can enter if you wish: irc://irc.gnome.org/gnome-design or https://matrix.to/#/#design:gnome.org). Let me copy some answers from GNOME designers to see if it rings any bells for you:

> "some self-baked header bar" is the proper solution

...

> does electron use gtk or is that a random x11 titlebar?
if the former, it's gtk-application-prefer-dark-theme
which will also make context menu etc dark
if the latter, gtk sets some atom, probably easier to look up in gtk source

...

> tldr gtk has an API for theme variants, if it's gtk you can use that. If it's not, check what atom gtk sets
and set it manually
I suppose either of that would need to be done on electron level


Also quoting from https://wiki.gnome.org/Initiatives/CSD:

> Many Electron apps should be low-hanging fruit, because they already use CSD on macOS (e.g. Slack). This likely means very little work would be required on the application side to make this happen on GNU/Linux as well.
> 
> However, since some additional complexity is required in order to get the proper GNOME CSD behavior (respecting gsettings etc.), some of this should probably be supported by Electron directly. That would be easier for Electron app developers and make for more consistent behavior.
> 
> We should first work with upstream Electron to add GNOME-style CSD support at a platform level. Once that is in place, we can start approaching individual app developers. 

Does Element use client-side decorations on mac?

**Comment by Exalm** [comment]
Looks like the Flathub version bundles a hacky script for doing this: https://github.com/flathub/com.spotify.Client/blob/master/set-dark-theme-variant.py

I have no idea what electron uses internally, but if it's GTK, you just need to flip [gtk-application-prefer-dark-theme](https://developer.gnome.org/gtk3/3.24/GtkSettings.html#GtkSettings--gtk-application-prefer-dark-theme). Note that it's not guaranteed the host theme can actually do this, so it may be ignored.

Otherwise, well, you need to manually set the x11 atom the gtk sets.

The best variant would be to use proper client-side decorations; the x11 approach will not work on Wayland anyway.

**Comment by SimonBrandner** [comment]
We are getting into an area in which I don't have experience...

As far as I understand this, if we can't do this using Electron, we could use a hack as the one @Exalm mentioned. But I don't want to, personally get into that, and I doubt that would get merged anyway...

**Comment by ZanderBrown** [comment]
I mean there is an extension for vscode that works fine, IIRC the implementation was a little hacky (and of course X11 only, but then electron is struck in the 00s anyway)

**Comment by jzvi12** [comment]
Hello guys, is there any updates on getting the tittle bar on dark mode? 

I'm new to the element app (it's awesome!), downloaded it for my Windows 10 PC; everything looks modern and elegant  expect for that title bar when in dark mode.

**Comment by SimonBrandner** [comment]
@jzvi12, I _think_ we can't do anything about this, especially on Windows. As far as I know, only apps with _custom_ title bars have theming in Windows, sorry...

**Comment by daenney** [comment]
Spotify is not an Electron app, it uses Chromium Embedded Framework directly. It's not a good proxy for what is isn't possible using Electron and hacks the Flathub might do for the Spotify app don't necessarily transpose to Electron apps.

The header bar in Gnome is controlled by your theme and if you switch you theme to a dark theme, it adapts just fine. This shouldn't be altered by the Element app, as it would be overriding system wide theme settings which is generally considered bad practice (even though some apps do it). In the case of the Spotify app it's probably done b/c the app is always in "dark mode" so it's less of an issue there to ignore the system theme.

**Comment by baek-sang** [comment]
Any update?

**Comment by claell** [comment]
VS Code (and other applications) do this.

On Windows, it looks like this for Element: 
![grafik](https://user-images.githubusercontent.com/6171135/169869706-f4f7ed5a-5333-4fd4-b9e1-e48b6018ce82.png)

**Comment by claell** [comment]
This might be a duplicate of #8749.

**Comment by t3chguy** [comment]
Duplicate of https://github.com/vector-im/element-web/issues/8749
